In [187]:
## Current path display

import os
current_path = os.getcwd()
print(current_path)

d:\05_Projekte\04_Research\08_ManuBrain\01_Lab\Demo-Cell-Classification\02_Preprocessing


In [188]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime
from datetime import timedelta  

## Functions 

In [189]:
# return tag from data
def fc_tag (data,index):
    return data.at[index,'d'].split('=')[2] # [2] equivalent a tag column data en txt file

# update tag in main table
def fc_tag_update(data, tag):
    for i in range(len(tag)):
        i = i + 1 
        data = data.replace({'PrimaryKey':i}, fc_tag(tag,i), regex = False)
    return data

#Sync TimeStamp

def fc_sync_timestamp2(data,ServerTimeStamp,delta_time): #Requires df, column-name and delta-time
    for date in data[ServerTimeStamp]:
        source = datetime.fromisoformat(str(date).replace('Z','+00:00'))
        date_sync = source - delta_time
        data[ServerTimeStamp] = date_sync
    return data

def fc_sync_timestamp(data,ServerTimeStamp,delta_time): #Requires df, column-name and delta-time
    for i in range(len(data)): 
        source = datetime.fromisoformat(str(data.at[i,ServerTimeStamp]).replace('Z','+00:00'))
        date_sync = source - delta_time
        data.at[i,ServerTimeStamp] = date_sync
    return data

## Paths from data set

In [190]:
date = 201030
uc_num = 6

path = str(date) + '/use_case' + str(uc_num)

# Conveyor_belt
path_data_conv = '../01_Data/' + path + '/Conveyor Speed/DataLogger.csv'
path_tag_conv = '../01_Data/' + path + '/Conveyor Speed/tagconfig.txt'

# Gate
path_data_gate = '../01_Data/' + path + '/Gate/DataLogger.csv'
path_tag_gate = '../01_Data/' + path + '/Gate/tagconfig.txt'

# Quality_Control
path_data_qc = '../01_Data/' + path + '/Quality Control/DataLogger.csv'
path_tag_qc = '../01_Data/' + path + '/Quality Control/tagconfig.txt'

# Robot_Position
path_data_rp = '../01_Data/' + path + '/Robot Position/DataLogger.csv'
path_tag_rp = '../01_Data/' + path + '/Robot Position/tagconfig.txt'

# Store
path_data_st = '../01_Data/' + path + '/Store/DataLogger.csv'
path_tag_st = '../01_Data/' + path + '/Store/tagconfig.txt'

# Flag
path_data_flag = '../01_Data/' + path + '/Flag/DataLogger.csv'
path_tag_flag = '../01_Data/' + path + '/Flag/tagconfig.txt'

# Processed data

path_data_save = '../01_Data/' + path + '/Processed/'
path_sample_save = '../01_Data/' + path + '/Processed/Samples/'


## Import data from csv

In [191]:
# Conveyor_belt
data_conv =  pd.read_csv(path_data_conv, delimiter = ',')
#df_conv = pd.DataFrame(pd.read_csv(path_data_conv, delimiter = ','))

# Gate
data_gate =  pd.read_csv(path_data_gate, delimiter = ',')
#df_gate = pd.DataFrame(pd.read_csv(path_data_gate, delimiter = ','))

# Quality_Control
data_qc =  pd.read_csv(path_data_qc, delimiter = ',')
#df_qc = pd.DataFrame(pd.read_csv(path_data_qc, delimiter = ','))

# Robot_Position
data_rp =  pd.read_csv(path_data_rp, delimiter = ',')
#df_rp = pd.DataFrame(pd.read_csv(path_data_rp, delimiter = ','))

# Store
data_st =  pd.read_csv(path_data_st, delimiter = ',')
#df_st = pd.DataFrame(pd.read_csv(path_data_st, delimiter = ','))

# Flag
data_flag =  pd.read_csv(path_data_flag, delimiter = ',')
#df_flag = pd.DataFrame(pd.read_csv(path_data_st, delimiter = ','))

## Import data tags

In [192]:
# Conveyor_belt
tag_conv =  pd.read_fwf(path_tag_conv, sep=" ", header=None, names=['a','b','c','d'])
tag_conv = tag_conv.drop([0]) # delete first row (trash)

# Gate
tag_gate =  pd.read_fwf(path_tag_gate, sep=" ", header=None, names=['a','b','c','d'])
tag_gate = tag_gate.drop([0]) # delete first row (trash)

# Quality_Control
tag_qc =  pd.read_fwf(path_tag_qc, sep=" ", header=None, names=['a','b','c','d'])
tag_qc = tag_qc.drop([0]) # delete first row (trash)

# Robot_Position
tag_rp =  pd.read_fwf(path_tag_rp, sep=" ", header=None, names=['a','b','c','d'])
tag_rp = tag_rp.drop([0]) # delete first row (trash)

# Robot_Position
tag_st =  pd.read_fwf(path_tag_st, sep=" ", header=None, names=['a','b','c','d'])
tag_st = tag_st.drop([0]) # delete first row (trash)

# Flag
tag_flag =  pd.read_fwf(path_tag_flag, sep=" ", header=None, names=['a','b','c','d'])
tag_flag = tag_flag.drop([0]) # delete first row (trash)

## Update tags

In [193]:
# Conveyor_belt
data_conv = fc_tag_update(data_conv, tag_conv)

# Gate
data_gate = fc_tag_update(data_gate, tag_gate)

# Quality_Control
data_qc = fc_tag_update(data_qc, tag_qc)

# Robot_Position
data_rp = fc_tag_update(data_rp, tag_rp)

# Store
data_st = fc_tag_update(data_st, tag_st)

# Flag
data_flag = fc_tag_update(data_flag, tag_flag)

## Syncronize ServerTimeStamp

In [194]:
delta_time_PLC = timedelta(seconds=67)
delta_time_SIM = timedelta(seconds=0)

In [195]:
data_conv = fc_sync_timestamp(data_conv,'ServerTimeStamp',delta_time_PLC)
data_conv

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0          27   0x0004    300  0x002f0000  2020-10-30T13:32:59.387Z   
1          28   0x0004    363  0x002f0000  2020-10-30T13:32:59.389Z   
2          29   0x0004    380  0x002f0000  2020-10-30T13:32:59.391Z   
3          27   0x0004    300  0x00000000  2020-10-30T13:32:59.658Z   
4          28   0x0004    363  0x00000000  2020-10-30T13:32:59.661Z   
5          29   0x0004    380  0x00000000  2020-10-30T13:32:59.663Z   
6          27   0x0004    300  0x002f0000  2020-10-30T13:34:03.696Z   
7          27   0x0004    300  0x00000000  2020-10-30T13:34:03.905Z   
8          28   0x0004    363  0x002f0000  2020-10-30T13:34:03.701Z   
9          28   0x0004    363  0x00000000  2020-10-30T13:34:03.908Z   
10         29   0x0004    380  0x002f0000  2020-10-30T13:34:03.706Z   
11         29   0x0004    380  0x00000000  2020-10-30T13:34:03.916Z   
12         27   0x0004    300  0x002f0000  2020-10-30T13:36:19.163Z   
13         27   0x0004    300  0x00000000  2020-10-30T13:36:19.376Z   
14         28   0x0004    363  0x002f0000  2020-10-30T13:36:19.168Z   
15         28   0x0004    363  0x00000000  2020-10-30T13:36:19.378Z   
16         29   0x0004    380  0x002f0000  2020-10-30T13:36:19.171Z   
17         29   0x0004    380  0x00000000  2020-10-30T13:36:19.378Z   
18         27   0x0004    300  0x002f0000  2020-10-30T13:36:32.374Z   
19         27   0x0004    300  0x00000000  2020-10-30T13:36:32.660Z   
20         28   0x0004    363  0x002f0000  2020-10-30T13:36:32.376Z   
21         28   0x0004    363  0x00000000  2020-10-30T13:36:32.663Z   
22         29   0x0004    380  0x002f0000  2020-10-30T13:36:32.376Z   
23         29   0x0004    380  0x00000000  2020-10-30T13:36:32.670Z   
24         27   0x0004    300  0x002f0000  2020-10-30T13:37:25.418Z   
25         27   0x0004    300  0x00000000  2020-10-30T13:37:25.633Z   
26         28   0x0004    363  0x002f0000  2020-10-30T13:37:25.420Z   
27         28   0x0004    363  0x00000000  2020-10-30T13:37:25.636Z   
28         29   0x0004    380  0x002f0000  2020-10-30T13:37:25.423Z   
29         29   0x0004    380  0x00000000  2020-10-30T13:37:25.638Z   
30         27   0x0004    300  0x002f0000  2020-10-30T13:40:34.376Z   
31         28   0x0004    363  0x002f0000  2020-10-30T13:40:34.377Z   
32         29   0x0004    380  0x002f0000  2020-10-30T13:40:34.379Z   
33         27   0x0004    300  0x00000000  2020-10-30T13:40:34.699Z   
34         28   0x0004    363  0x00000000  2020-10-30T13:40:34.704Z   
35         29   0x0004    380  0x00000000  2020-10-30T13:40:34.714Z   
36         27   0x0004    300  0x002f0000  2020-10-30T13:40:51.123Z   
37         28   0x0004    363  0x002f0000  2020-10-30T13:40:51.126Z   
38         29   0x0004    380  0x002f0000  2020-10-30T13:40:51.128Z   
39         27   0x0004    300  0x00000000  2020-10-30T13:40:51.416Z   
40         28   0x0004    363  0x00000000  2020-10-30T13:40:51.421Z   
41         29   0x0004    380  0x00000000  2020-10-30T13:40:51.429Z   

                     ServerTimeStamp  
0   2020-10-30 13:31:52.387000+00:00  
1   2020-10-30 13:31:52.389000+00:00  
2   2020-10-30 13:31:52.391000+00:00  
3   2020-10-30 13:31:52.658000+00:00  
4   2020-10-30 13:31:52.661000+00:00  
5   2020-10-30 13:31:52.663000+00:00  
6   2020-10-30 13:32:56.696000+00:00  
7   2020-10-30 13:32:56.905000+00:00  
8   2020-10-30 13:32:56.701000+00:00  
9   2020-10-30 13:32:56.908000+00:00  
10  2020-10-30 13:32:56.706000+00:00  
11  2020-10-30 13:32:56.916000+00:00  
12  2020-10-30 13:35:12.163000+00:00  
13  2020-10-30 13:35:12.376000+00:00  
14  2020-10-30 13:35:12.168000+00:00  
15  2020-10-30 13:35:12.378000+00:00  
16  2020-10-30 13:35:12.171000+00:00  
17  2020-10-30 13:35:12.378000+00:00  
18  2020-10-30 13:35:25.374000+00:00  
19  2020-10-30 13:35:25.660000+00:00  
20  2020-10-30 13:35:25.376000+00:00  
21  2020-10-30 13:35:25.663000+00:00  
22  2020-10-30 13:35:25.376000+00:00  
23  2020-10-3

In [196]:
data_gate = fc_sync_timestamp(data_gate,'ServerTimeStamp',delta_time_SIM)
data_gate

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0       Position   0x000b    0.3  0x00000000  2020-10-30T13:19:18.888Z   
1       Position   0x000b    0.3  0x00000000  2020-10-30T13:19:19.188Z   
2           Safe   0x0001   true  0x00000000  2020-10-30T13:19:18.888Z   
3           Safe   0x0001   true  0x00000000  2020-10-30T13:19:19.188Z   
4          Speed   0x000b      0  0x00000000  2020-10-30T13:19:18.888Z   
...          ...      ...    ...         ...                       ...   
18328   Position   0x000b     69  0x00000000  2020-10-30T13:49:51.888Z   
18329       Safe   0x0001  false  0x00000000  2020-10-30T13:49:51.588Z   
18330       Safe   0x0001  false  0x00000000  2020-10-30T13:49:51.888Z   
18331      Speed   0x000b    -40  0x00000000  2020-10-30T13:49:51.588Z   
18332      Speed   0x000b    -30  0x00000000  2020-10-30T13:49:51.888Z   

                        ServerTimeStamp  
0      2020-10-30 13:19:18.888000+00:00  
1      2020-10-30 13:19:19.188000+00:00  
2      2020-10-30 13:19:18.888000+00:00  
3      2020-10-30 13:19:19.188000+00:00  
4      2020-10-30 13:19:18.888000+00:00  
...                                 ...  
18328  2020-10-30 13:49:51.888000+00:00  
18329  2020-10-30 13:49:51.588000+00:00  
18330  2020-10-30 13:49:51.888000+00:00  
18331  2020-10-30 13:49:51.588000+00:00  
18332  2020-10-30 13:49:51.888000+00:00  

[18333 rows x 6 columns]

In [197]:
data_qc = fc_sync_timestamp(data_qc,'ServerTimeStamp',delta_time_SIM)
data_qc

PrimaryKey DataType     Value  StatusCode           SourceTimeStamp  \
0      DistanceAbs   0x000b  6.100943  0x00000000  2020-10-30T13:19:19.903Z   
1      DistanceAbs   0x000b  6.100943  0x00000000  2020-10-30T13:19:20.203Z   
2        DistanceX   0x000b  5.422691  0x00000000  2020-10-30T13:19:19.903Z   
3        DistanceX   0x000b  5.422691  0x00000000  2020-10-30T13:19:20.203Z   
4        DistanceY   0x000b  2.795699  0x00000000  2020-10-30T13:19:19.903Z   
...            ...      ...       ...         ...                       ...   
24439   ResultCode   0x0006  1.000000  0x00000000  2020-10-30T13:49:52.604Z   
24440  DistanceAbs   0x000b  5.274006  0x00000000  2020-10-30T13:49:52.903Z   
24441    DistanceX   0x000b  1.084538  0x00000000  2020-10-30T13:49:52.903Z   
24442    DistanceY   0x000b  5.161290  0x00000000  2020-10-30T13:49:52.903Z   
24443   ResultCode   0x0006  1.000000  0x00000000  2020-10-30T13:49:52.904Z   

                        ServerTimeStamp  
0      2020-10-30 13:19:19.903000+00:00  
1      2020-10-30 13:19:20.203000+00:00  
2      2020-10-30 13:19:19.903000+00:00  
3      2020-10-30 13:19:20.203000+00:00  
4      2020-10-30 13:19:19.903000+00:00  
...                                 ...  
24439  2020-10-30 13:49:52.604000+00:00  
24440  2020-10-30 13:49:52.903000+00:00  
24441  2020-10-30 13:49:52.903000+00:00  
24442  2020-10-30 13:49:52.903000+00:00  
24443  2020-10-30 13:49:52.904000+00:00  

[24444 rows x 6 columns]

In [198]:
data_rp = fc_sync_timestamp(data_rp,'ServerTimeStamp',delta_time_PLC)
data_rp

PrimaryKey DataType       Value  StatusCode           SourceTimeStamp  \
0            527   0x0007  1051168136  0x00000000  2020-10-30T13:20:31.122Z   
1            528   0x0007  3188582587  0x00000000  2020-10-30T13:20:31.124Z   
2            529   0x0007  1051728003  0x00000000  2020-10-30T13:20:31.127Z   
3            527   0x0007  1051169144  0x00000000  2020-10-30T13:20:31.373Z   
4            527   0x0007  1051169176  0x00000000  2020-10-30T13:20:31.620Z   
...          ...      ...         ...         ...                       ...   
21373        527   0x0007  3180520831  0x00000000  2020-10-30T13:51:03.624Z   
21374        528   0x0007  3190996050  0x00000000  2020-10-30T13:51:03.397Z   
21375        528   0x0007  3190581969  0x00000000  2020-10-30T13:51:03.627Z   
21376        529   0x0007  1054427313  0x00000000  2020-10-30T13:51:03.400Z   
21377        529   0x0007  1054422475  0x00000000  2020-10-30T13:51:03.627Z   

                        ServerTimeStamp  
0      2020-10-30 13:19:24.122000+00:00  
1      2020-10-30 13:19:24.124000+00:00  
2      2020-10-30 13:19:24.127000+00:00  
3      2020-10-30 13:19:24.373000+00:00  
4      2020-10-30 13:19:24.620000+00:00  
...                                 ...  
21373  2020-10-30 13:49:56.624000+00:00  
21374  2020-10-30 13:49:56.397000+00:00  
21375  2020-10-30 13:49:56.627000+00:00  
21376  2020-10-30 13:49:56.400000+00:00  
21377  2020-10-30 13:49:56.627000+00:00  

[21378 rows x 6 columns]

In [199]:
data_st = fc_sync_timestamp(data_st,'ServerTimeStamp',delta_time_SIM)
data_st

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0      DiscSize   0x0003      1  0x00000000  2020-10-30T13:19:22.202Z   
1      DiscSize   0x0003      1  0x00000000  2020-10-30T13:19:22.502Z   
2      DiscSize   0x0003      1  0x00000000  2020-10-30T13:19:22.802Z   
3      DiscSize   0x0003      1  0x00000000  2020-10-30T13:19:23.102Z   
4      DiscSize   0x0003      1  0x00000000  2020-10-30T13:19:23.402Z   
...         ...      ...    ...         ...                       ...   
6106   DiscSize   0x0003      2  0x00000000  2020-10-30T13:49:54.002Z   
6107   DiscSize   0x0003      2  0x00000000  2020-10-30T13:49:54.302Z   
6108   DiscSize   0x0003      2  0x00000000  2020-10-30T13:49:54.602Z   
6109   DiscSize   0x0003      2  0x00000000  2020-10-30T13:49:54.902Z   
6110   DiscSize   0x0003      2  0x00000000  2020-10-30T13:49:55.202Z   

                       ServerTimeStamp  
0     2020-10-30 13:19:22.202000+00:00  
1     2020-10-30 13:19:22.502000+00:00  
2     2020-10-30 13:19:22.802000+00:00  
3     2020-10-30 13:19:23.102000+00:00  
4     2020-10-30 13:19:23.402000+00:00  
...                                ...  
6106  2020-10-30 13:49:54.002000+00:00  
6107  2020-10-30 13:49:54.302000+00:00  
6108  2020-10-30 13:49:54.602000+00:00  
6109  2020-10-30 13:49:54.902000+00:00  
6110  2020-10-30 13:49:55.202000+00:00  

[6111 rows x 6 columns]

In [200]:
data_flag = fc_sync_timestamp(data_flag,'ServerTimeStamp',delta_time_PLC)
data_flag

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0          561   0x0001   true  0x00000000  2020-10-30T13:20:26.618Z   
1          560   0x0004      2  0x00000000  2020-10-30T13:20:26.621Z   
2          560   0x0004      3  0x00000000  2020-10-30T13:20:31.119Z   
3          561   0x0001  false  0x00000000  2020-10-30T13:20:31.871Z   
4          560   0x0004      0  0x00000000  2020-10-30T13:20:36.871Z   
..         ...      ...    ...         ...                       ...   
555        561   0x0001   true  0x00000000  2020-10-30T13:50:48.619Z   
556        560   0x0004      3  0x00000000  2020-10-30T13:50:53.372Z   
557        561   0x0001  false  0x00000000  2020-10-30T13:50:54.119Z   
558        560   0x0004      0  0x00000000  2020-10-30T13:50:59.195Z   
559        561   0x0001   true  0x00000000  2020-10-30T13:51:00.126Z   

                      ServerTimeStamp  
0    2020-10-30 13:19:19.618000+00:00  
1    2020-10-30 13:19:19.621000+00:00  
2    2020-10-30 13:19:24.119000+00:00  
3    2020-10-30 13:19:24.871000+00:00  
4    2020-10-30 13:19:29.871000+00:00  
..                                ...  
555  2020-10-30 13:49:41.619000+00:00  
556  2020-10-30 13:49:46.372000+00:00  
557  2020-10-30 13:49:47.119000+00:00  
558  2020-10-30 13:49:52.195000+00:00  
559  2020-10-30 13:49:53.126000+00:00  

[560 rows x 6 columns]

## Merge Data Sets

In [201]:
data_set = [data_conv, data_gate, data_qc, data_rp, data_st, data_flag]
data_merge = pd.concat(data_set, keys=['conv', 'gate', 'qc', 'rp', 'store', 'flag'])
# Sort the data_set by SourceTimeStamp, reset index and delete aux column
data_merge = data_merge.sort_values(by='ServerTimeStamp').reset_index().drop('level_1',1)
#del data_merge['level_1'] #delete aux column

In [202]:
data_merge

level_0 PrimaryKey DataType       Value  StatusCode  \
0        gate       Safe   0x0001        true  0x00000000   
1        gate   Position   0x000b         0.3  0x00000000   
2        gate      Speed   0x000b           0  0x00000000   
3        gate      Speed   0x000b           0  0x00000000   
4        gate       Safe   0x0001        true  0x00000000   
...       ...        ...      ...         ...         ...   
70863      rp        528   0x0007  3190996050  0x00000000   
70864      rp        529   0x0007  1054427313  0x00000000   
70865      rp        527   0x0007  3180520831  0x00000000   
70866      rp        528   0x0007  3190581969  0x00000000   
70867      rp        529   0x0007  1054422475  0x00000000   

                SourceTimeStamp                   ServerTimeStamp  
0      2020-10-30T13:19:18.888Z  2020-10-30 13:19:18.888000+00:00  
1      2020-10-30T13:19:18.888Z  2020-10-30 13:19:18.888000+00:00  
2      2020-10-30T13:19:18.888Z  2020-10-30 13:19:18.888000+00:00  
3      2020-10-30T13:19:19.188Z  2020-10-30 13:19:19.188000+00:00  
4      2020-10-30T13:19:19.188Z  2020-10-30 13:19:19.188000+00:00  
...                         ...                               ...  
70863  2020-10-30T13:51:03.397Z  2020-10-30 13:49:56.397000+00:00  
70864  2020-10-30T13:51:03.400Z  2020-10-30 13:49:56.400000+00:00  
70865  2020-10-30T13:51:03.624Z  2020-10-30 13:49:56.624000+00:00  
70866  2020-10-30T13:51:03.627Z  2020-10-30 13:49:56.627000+00:00  
70867  2020-10-30T13:51:03.627Z  2020-10-30 13:49:56.627000+00:00  

[70868 rows x 7 columns]

## Export Data to csv

In [203]:
data_merge.to_csv(path_data_save + 'data_merge.csv')
print('The data is saved at ' + path_data_save + 'data_merge.csv')

The data is saved at ../01_Data/201030/use_case6/Processed/data_merge.csv


In [204]:
len(data_merge)

70868

## Determine the starting point of data set

#Based on change of DistanceAbs in Data- Set
var_actual = 0
run_list = [0]
for i in range(len(data_merge)):
    var_previous = var_actual
    if data_merge.at[i,'PrimaryKey'] == 'DistanceAbs':  # if 'PrimaryKey' value = 'DistanceAbs':
        var_actual = data_merge.at[i,'Value'] # value of current DistanceAbs
    if var_previous != var_actual:
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample    

# Based on grabbing the white disk
var_actual = 99
disk_actual = 99 # Initial value
run_list = [0]
disk = False
storage_down = False
rob_x = 0
rob_y = 0
rob_z = 0
rob_pos = 0
pos = [0]*(len(data_merge)-2)
for i in range(len(data_merge)-2): # len(data_merge)-2 is considered to avoid border errors
    #var_previous = var_actual
    #rob_pos = fc_rob_pos(i, data_merge) # review robot position
    pos[i] = 0
    if data_merge.at[i,'PrimaryKey'] == 'DiscSize':  # if 'PrimaryKey' value = 'DiskSize':
        disk_previous = disk_actual
        disk_actual = data_merge.at[i,'Value'] # value of current DiskSize
        #print(disk_previous, disk_actual, i)
        if disk_previous != 2 and disk_actual == 2: # Varialbe 0: unkown; 1: Black disk; 2: white disk
            disk = True
            #print('New white disk here ' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '527':  # if 'PrimaryKey' value = 'x coordinate':
        rob_x = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('x = ' + str(rob_x) + '--' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '528':  # if 'PrimaryKey' value = 'y coordinate':
        rob_y = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('y = ' + str(rob_y) + '--' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '529':  # if 'PrimaryKey' value = 'z coordinate':
        rob_z = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('z = ' + str(rob_z) + '--' + str(i))
    if rob_x > 3100000000 and rob_y > 3100000000 and rob_z < 1051200000: #Robot in storage. Picking up the disk
        rob_pos = 1
        pos[i] = 1
        #print('pos 1 - ' + str(i))
    if rob_x < 1052000000 and rob_y > 3100000000 and rob_z < 1052000000 and rob_z > 1051200000 : #Robot in storage. Placing down the disk
        rob_pos = 2
        pos[i] = 2
        #print('pos 2 - ' + str(i))
    #print(rob_pos)
    if rob_pos == 1: # Robot picking up a disk
        storage_down = True
        #print(storage_down,i)  
    if disk == True and storage_down == True:
        #var_actual = rob_pos
        disk = False
        storage_down = False
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample
        #print(i)
    #if var_previous != var_actual:
        #run_list.append(i) #i append to list # This list will tell at how many i contains one sample   
 

In [205]:
# Based on grabbing the white disk
run_list = [0]
rob_pos_actual = 99
rob_pos_old = 0
gripper_actual = 99
gripper_old = 0
gripper = False
rob_pos = False
for i in range(len(data_merge)-2): # len(data_merge)-2 is considered to avoid border errors
    #var_previous = var_actual
    #rob_pos = fc_rob_pos(i, data_merge) # review robot position
    if data_merge.at[i,'PrimaryKey'] == '561':  # if 'PrimaryKey' value = 'gripper open (0), close (1)':
        gripper_old = gripper_actual
        gripper_actual = data_merge.at[i,'Value'] # value of current 'x coordinate'
        if gripper_old != gripper_actual:
            gripper = True
    if data_merge.at[i,'PrimaryKey'] == '560':  # if 'PrimaryKey' value = 'position of robot based in certain conditions':
        # 0: Black disk in store
        # 1: Black disk in gate
        # 2: White disk in store
        # 3: White disk in gate
        rob_pos_old = rob_pos_actual
        rob_pos_actual = int(data_merge.at[i,'Value']) # value of current 'x coordinate'
        if rob_pos_old != rob_pos_actual and rob_pos_actual == 2:
            rob_pos = True
    if rob_pos == True and gripper == True:
        rob_pos = False
        gripper = False
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample
        #print(i)
 

In [206]:
print(run_list) # Starting points. 
print(len(run_list))

[0, 10, 883, 1902, 2929, 3948, 4975, 6532, 7555, 8567, 10128, 11151, 12724, 13747, 14781, 15793, 16823, 17846, 18862, 19881, 20908, 21927, 22954, 23973, 24999, 26017, 27044, 28063, 29097, 30116, 31143, 32172, 33203, 34226, 35249, 36272, 37308, 38331, 39360, 40383, 41407, 42429, 43452, 44472, 45499, 46521, 47556, 48578, 49602, 50624, 51646, 52658, 53692, 54715, 55738, 57327, 58358, 60041, 61071, 62098, 63117, 64144, 65174, 66201, 67231, 68254, 69278, 70300]
68


In [207]:
# data_merge.iloc[list(range(0,3))]

In [208]:
#df_robpos = pd.DataFrame(pos,columns=['RobPos']) # Position of robot. 0: not important; 1: Robot in storage. Picking up the disk; 2:Robot in storage. Placing down the disk
#data_merge = pd.concat([data_merge, df_robpos], axis=1, sort=False)
data_merge

level_0 PrimaryKey DataType       Value  StatusCode  \
0        gate       Safe   0x0001        true  0x00000000   
1        gate   Position   0x000b         0.3  0x00000000   
2        gate      Speed   0x000b           0  0x00000000   
3        gate      Speed   0x000b           0  0x00000000   
4        gate       Safe   0x0001        true  0x00000000   
...       ...        ...      ...         ...         ...   
70863      rp        528   0x0007  3190996050  0x00000000   
70864      rp        529   0x0007  1054427313  0x00000000   
70865      rp        527   0x0007  3180520831  0x00000000   
70866      rp        528   0x0007  3190581969  0x00000000   
70867      rp        529   0x0007  1054422475  0x00000000   

                SourceTimeStamp                   ServerTimeStamp  
0      2020-10-30T13:19:18.888Z  2020-10-30 13:19:18.888000+00:00  
1      2020-10-30T13:19:18.888Z  2020-10-30 13:19:18.888000+00:00  
2      2020-10-30T13:19:18.888Z  2020-10-30 13:19:18.888000+00:00  
3      2020-10-30T13:19:19.188Z  2020-10-30 13:19:19.188000+00:00  
4      2020-10-30T13:19:19.188Z  2020-10-30 13:19:19.188000+00:00  
...                         ...                               ...  
70863  2020-10-30T13:51:03.397Z  2020-10-30 13:49:56.397000+00:00  
70864  2020-10-30T13:51:03.400Z  2020-10-30 13:49:56.400000+00:00  
70865  2020-10-30T13:51:03.624Z  2020-10-30 13:49:56.624000+00:00  
70866  2020-10-30T13:51:03.627Z  2020-10-30 13:49:56.627000+00:00  
70867  2020-10-30T13:51:03.627Z  2020-10-30 13:49:56.627000+00:00  

[70868 rows x 7 columns]

The list contains the point i when the value is different. To consider the point where the value remains, consider i-1

In [209]:
d = {} # Dictionary that holds the sub-data-sets given by the starting point
for i in range(len(run_list)-1): # Ignores the last value (set of values)
    list_range = list(range(run_list[i],run_list[i+1]))
    d['df{0}'.format(i)] = data_merge.iloc[list_range]
    
print('Dictionary lenght is:' + str(len(d)))

Dictionary lenght is:67


## Assure conv_speed per sample 

In [210]:
conv_1 = 0
conv_2 = 0
conv_3 = 0
conv1 = False
conv2 = False
conv3 = False
for j in range(len(d)): # as many samples saved in dictionary
    df = d['df'+ str(j)]
    for i in range(len(df)): # as many row in the dataFrame
        #print(df.iat[i,1])
        if df.iat[i,1] == '27': # conv1
            conv_1 = df.iat[i,3]
            conv1 = True # the value shows at least once in the data set
        if df.iat[i,1] == '28': # conv2
            conv_2 = df.iat[i,3]
            conv2 = True
        if df.iat[i,1] == '29': # conv3
            conv_3 = df.iat[i,3]
            conv3 = True
    if conv1 == False: # The value did not occur in this sample. Copy value from previous samples into this one
        dic = {'PrimaryKey' : '27', 'Value': str(conv_1)}
        df.loc[len(df)+1,'PrimaryKey'] = '27'
        df.loc[len(df),'Value'] = str(conv_1)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
    if conv2 == False:
        dic = {'PrimaryKey' : '27', 'Value': str(conv_2)}
        df.loc[len(df)+1,'PrimaryKey'] = '28'
        df.loc[len(df),'Value'] = str(conv_2)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
    if conv3 == False:
        dic = {'PrimaryKey' : '27', 'Value': str(conv_3)}
        df.loc[len(df)+1,'PrimaryKey'] = '29'
        df.loc[len(df),'Value'] = str(conv_3)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
   
    conv1 = False # reset conv for the next round
    conv2 = False
    conv3 = False
print(conv_1)
print(conv_2)
print(conv_3)

300
363
380


## Export individual data-sample to .csv

In [211]:
for i in range(len(d)):
    if i > 0: # always ignore the first sample due to data-acquisition misalligment
        d['df' + str(i)].to_csv(path_sample_save + 'sample_' + str(i) + '.csv')
print('The data is saved at ' + path_sample_save + 'sample_x.csv')

The data is saved at ../01_Data/201030/use_case6/Processed/Samples/sample_x.csv


## Testing area

In [212]:
#d['df67']

In [213]:
# list_range = list(range(run_list[0],run_list[1]))
# print(list_range)

In [214]:
# data_merge.iloc[list_range]

In [215]:
# data_rp.loc[:, 'PrimaryKey'] = '529_z'
# data_rp

In [216]:
# %matplotlib notebook 
# data_gate.head(400).plot(x='SourceTimeStamp', y= 'Value')
# plt.xticks(rotation=45)

In [217]:
# %matplotlib notebook 
# data_rp.head(200).plot( y= 'Value')
# plt.xticks(rotation=45)